veros/core/density/gsw.py

```diff
@@ -545,3 +545,489 @@ def gsw_dHdS(vs, sa_in, ct_in, p):
+           y*(674.819060538734 + z*(-1069.887337245828 + (530.4484299696 - 158.40030944233638*z)*z) + \
+             y*(-409.779283929806 + y*(149.452282277512 - 218.92375140095282*z) + \
+              (681.370187043564 - 133.5392811916956*z)*z))))
+
+      return (g03 + g08) * 1e-16
+
+    else:
+
+      raise ValueError("nsd, ntd and npd values do not match (given: {}, {}, {})".\
+            format(nsd, ntd, npd))
+
+@veros_method
+def gsw_rho_t_exact(vs, sa, t, p):
+    """
+    Calculates in-situ density of seawater from Absolute Salinity and
+    in-situ temperature.
+
+    Arguments:
+        sa     : Absolute Salinity                               [ g/kg ]
+        t      : in-situ temperature                              [deg C]
+        p      : sea pressure                                      [dbar]
+
+    Returns:
+        in-situ density                                          [kg/m^3]
+    """
+
+    #N0=0
+    #N1=1
+
+    return 1.0 / gsw_gibbs(vs, 0, 0, 1, sa, t, p)
+
+@veros_method
+def gsw_pot_rho_t_exact(vs, sa, t, p, p_ref):
+    """
+    Calculates the potential density of seawater
+
+    Arguments:
+        sa     : Absolute Salinity                               [ g/kg ]
+        t      : in-situ temperature                              [deg C]
+        p      : sea pressure                                      [dbar]
+        p_ref  : reference sea pressure                            [dbar]
+
+    Returns:
+        Potential density                                        [kg/m^3]
+    """
+
+    pt = gsw_pt_from_t(vs, sa, t, p, p_ref)
+    return gsw_rho_t_exact(vs, sa, pt, p_ref)
+
+@veros_method
+def gsw_pt_from_t(vs, sa, t, p, p_ref):
+    """
+    Calculates potential temperature of seawater from in-situ temperature
+
+    Arguments:
+        sa     : Absolute Salinity                               [ g/kg ]
+        t      : in-situ temperature                              [deg C]
+        p      : sea pressure                                      [dbar]
+        p_ref  : reference sea pressure                            [dbar]
+
+    Returns:
+        Potential temperature                                     [deg C]
+    """
+
+    #N0 = 0
+    #N2 = 2
+
+    s1 = sa / get_gsw_teos10_constants(vs, 'gsw_ups')
+
+    pt = t + (p-p_ref)*( 8.65483913395442e-6 - \
+                   s1 *  1.41636299744881e-6 - \
+             (p+p_ref)*  7.38286467135737e-9 + \
+                   t  *(-8.38241357039698e-6 + \
+                   s1 *  2.83933368585534e-8 + \
+                   t  *  1.77803965218656e-8 + \
+             (p+p_ref)*  1.71155619208233e-10))
+
+    dentropy_dt = get_gsw_teos10_constants(vs, 'gsw_cp0') /\
+                  ((get_gsw_teos10_constants(vs, 'gsw_t0') + pt) *\
+                  (1.0 - 0.05 * (1.0 - sa / get_gsw_teos10_constants(vs, 'gsw_sso'))))
+
+    true_entropy_part = gsw_entropy_part(vs, sa, t, p)
+
+    for _ in range(2):
```

Comment:
This is very confusing. Why do it twice? This needs at least a comment, and should be re-formulated if possible.

Reply:
This routine was removed since the method for potential density computation has been changed. I was using TEOS-10 routines, which have input parameter of in-situ temperature but  Veros's temperature variable is the "conservative" one.